### PR TITLE
Use Response Files in CMake

### PR DESCRIPTION
**Pull request by jmittert**
Passing all of the swift source files to the swift compiler causes the
command line to go over Window's 32767 character limit command length.

This now writes the source file lists to response files to get around that.

I'm a CMake newb (it took me far too long to figure out how to join a list into a string), so it's fairly likely there's a better way of doing this.

### CODE REVIEW

File: cmake/modules/SwiftSupport.cmake [hunk 1]
```diff
@@ -97,6 +97,10 @@ function(add_swift_target target)
 
     set(all_sources ${sources})
     list(INSERT all_sources ${i} -primary-file)
+    string(LENGTH all_sources source_length)
+    set(resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/${name}.resp)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: Traditionally, the extension is `.rsp`.

File: cmake/modules/SwiftSupport.cmake [hunk 2]
```diff
@@ -106,11 +110,22 @@ function(add_swift_target target)
...
+    list(JOIN objs " " objs_resp_text)
+    file(WRITE ${objs_resp_file} ${objs_resp_text})
+
     list(APPEND mods ${mod})
+    set(mods_resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/modects.resp)
+    list(JOIN mods " " mods_resp_text)
+    file(WRITE ${mods_resp_file} ${mods_resp_text})
+
     list(APPEND docs ${doc})
+    set(docs_resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/docsects.resp)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: Similar

File: cmake/modules/SwiftSupport.cmake [hunk 3]
```diff
@@ -106,11 +110,22 @@ function(add_swift_target target)
...
-                         ${CMAKE_SWIFT_COMPILER} -frontend ${compile_flags} -emit-module-path ${mod} -emit-module-doc-path ${doc} -o ${obj} -c ${all_sources})
+                         ${CMAKE_SWIFT_COMPILER} -frontend ${compile_flags} -emit-module-path ${mod} -emit-module-doc-path ${doc} -o ${obj} -c "@${resp_file}")
 
     list(APPEND objs ${obj})
+    set(objs_resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/objects.resp)
+    list(JOIN objs " " objs_resp_text)
+    file(WRITE ${objs_resp_file} ${objs_resp_text})
+
     list(APPEND mods ${mod})
+    set(mods_resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/modects.resp)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: Similar.  Also, did you mean modsects?  Whats the `ects` suffix supposed to stand for?

File: cmake/modules/SwiftSupport.cmake [hunk 4]
```diff
@@ -106,11 +110,22 @@ function(add_swift_target target)
                          ${source}
                          ${AST_DEPENDS}
                        COMMAND
-                         ${CMAKE_SWIFT_COMPILER} -frontend ${compile_flags} -emit-module-path ${mod} -emit-module-doc-path ${doc} -o ${obj} -c ${all_sources})
+                         ${CMAKE_SWIFT_COMPILER} -frontend ${compile_flags} -emit-module-path ${mod} -emit-module-doc-path ${doc} -o ${obj} -c "@${resp_file}")
 
     list(APPEND objs ${obj})
+    set(objs_resp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/objects.resp)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: Similar

File: cmake/modules/SwiftSupport.cmake (hunk 3 above)
Comment (jmittert) [reply cmake/modules/SwiftSupport.cmake]: eee, should just be mods, bad search and replace after copying it from the objects one.

**Comment by jmittert** [comment]
I can make this change in dispatch and xctest as well if that's wanted. However, what would be the better way of doing it?

**Comment by jmittert** [comment]
Alright, updated this to use the swiftc driver + output_file_maps to track dependencies. The initial Windows build will be slower until we get a multithreaded swiftc Windows driver in, but the once built it's quick to rebuild.

File: cmake/modules/SwiftSupport.cmake [hunk 5]
```diff
@@ -1,6 +1,36 @@
...
+    set(dia ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/${name}.dia)
+    set(output_entry "\"${source}\": {\n\
+    \"object\": \"${obj}\",\n\
+    \"dependencies\": \"${deps}\",\n\
+    \"swift-dependencies\": \"${swiftdeps}\",\n\
+    \"diagnostics\": \"${dia}\"\n\
+}")
+    list(APPEND output_list ${output_entry})
+  endforeach()
+  set(master_deps ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/main-build-record.swiftdeps)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: I think that the name should be `${target}.swiftdeps`.

File: cmake/modules/SwiftSupport.cmake [hunk 6]
```diff
@@ -84,53 +115,14 @@ function(add_swift_target target)
...
-
-    math(EXPR i "${i}+1")
-  endforeach()
-
-  if(AST_LIBRARY)
-    get_filename_component(module_directory ${AST_MODULE_PATH} DIRECTORY)
-
-    set(module ${AST_MODULE_PATH})
-    set(documentation ${module_directory}/${AST_MODULE_NAME}.swiftdoc)
+  set(output_map_path ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/output_file_map.json)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: I'd prefer `output-file-map.json`.

File: cmake/modules/SwiftSupport.cmake [hunk 7]
```diff
@@ -84,53 +115,14 @@ function(add_swift_target target)
...
-                         ${documentation}
-                       DEPENDS
-                         ${mods}
-                         ${docs}
-                         ${AST_DEPENDS}
-                       COMMAND
-                         ${CMAKE_SWIFT_COMPILER} -frontend ${compile_flags} -sil-merge-partial-modules -emit-module ${mods} -o ${module} -emit-module-doc-path ${documentation})
-  endif()
+  string(LENGTH sources source_length)
+  set(rsp_file ${CMAKE_CURRENT_BINARY_DIR}/${target}.dir/${target}.rsp)
```
Comment (compnerd) [inline cmake/modules/SwiftSupport.cmake]: I think that we should be able to drop the `rsp` file here and let CMake/ninja DTRT.

Comment (jmittert) [reply cmake/modules/SwiftSupport.cmake]: DTRT?

Comment (jmittert) [reply cmake/modules/SwiftSupport.cmake]: Oh, do the right thing. CMake/Ninja don't do the right thing though. They write a batch file, but that's still over the 32k character limit for CreateProcessW.

Comment (compnerd) [reply cmake/modules/SwiftSupport.cmake]: Ugh, right, because of the `add_custom_command` rather than the `add_library` *sigh*

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@jmittert:

20:02:09 CMake Error at cmake/modules/SwiftSupport.cmake:29 (list):
20:02:09   list does not recognize sub-command JOIN

**Comment by jmittert** [comment]
Oh.. list(JOIN...) is cmake 3.12 I'll find an alternative

**Comment by jmittert** [comment]
Should be good now. Switched it to not use an intermediate list.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@compnerd is this good to go?

**Comment by compnerd** [comment]
Yeap, this is good to go!  Thanks for the ping (wow, so that's what its like on the other side :-p)

**Pull request merged**